phase1-coordinator/src/main.rs

```diff
@@ -26,34 +29,51 @@ pub async fn main() -> anyhow::Result<()> {
     // let environment: Environment = Production::from(Parameters::AleoInner).into();
 
     // Instantiate the coordinator.
-    let coordinator = coordinator(&environment, Box::new(Dummy)).await?;
+    let coordinator: Arc<RwLock<Coordinator<Disk>>> =
+        Arc::new(RwLock::new(coordinator(&environment, Arc::new(Dummy))?));
 
+    let ceremony_coordinator = coordinator.clone();
     // Initialize the coordinator.
-    let operator = coordinator.clone();
     let ceremony = task::spawn(async move {
         // Initialize the coordinator.
-        operator.initialize().unwrap();
+        ceremony_coordinator.write().await.initialize().unwrap();
 
         // Initialize the coordinator loop.
         loop {
             // Run the update operation.
-            if let Err(error) = operator.update() {
+            if let Err(error) = ceremony_coordinator.write().await.update() {
                 error!("{}", error);
             }
 
             // Sleep for 10 seconds in between iterations.
             sleep(Duration::from_secs(10)).await;
+            break;
         }
     });
 
     // Initialize the shutdown procedure.
-    let handler = coordinator.clone();
-    {
-        debug!("Initializing the shutdown handler");
-        handler.shutdown_listener()?;
-    }
+    let shutdown_handler = {
+        let shutdown_coordinator = coordinator.clone();
+        task::spawn(async move {
+            tokio::signal::ctrl_c()
+                .await
+                .expect("Error while waiting for shutdown signal");
+            shutdown_coordinator
+                .write()
+                .await
+                .shutdown()
+                .expect("Error while shutting down");
+        })
+    };
 
-    ceremony.await.expect("The ceremony handle has panicked");
+    tokio::select! {
+        _ = shutdown_handler => {
+            println!("Shutdown completed first")
+        }
+        _ = ceremony => {
+            println!("Ceremony completed first")
+        }
+    };
```

Comment:
In docs for `select!`
> Wait on multiple concurrent branches, returning when the first branch completes, cancelling the remaining branches.

I don't know how often do we plan to kill the process, but it's a potential source of bugs. The futures in select should be cancellation-safe. So far we did nothing to ensure coordinator has this quality.